onnx_caffe2/backend.py

```diff
@@ -297,7 +300,71 @@ def _create_gemm(cls, n, opset_version):
         return ops
 
     @classmethod
-    def _create_pad(cls, n, opset_version):
+    def _create_lstm(cls, onnx_model, n, opset_version):
+        assert onnx_model is not None, "cannot convert LSTMs through this codepath"
+
+        attrs = dict(n.attrs)
+        hidden_size = attrs.pop('hidden_size')
+        assert not attrs, "unsupported LSTM attributes: " + attrs.keys()
+
+        input_blob, W, R, B, sequence_lens, initial_h, initial_c = n.inputs
+
+        # ad-hoc, informally-specified, bug-ridden, slow
+        # implementation of shape inference
+        for x in onnx_model.graph.input:
+            if x.name == input_blob:
+                input_size = x.type.tensor_type.shape.dim[2].dim_value
+        for x in map(OnnxNode, onnx_model.graph.node):
+            if x.op_type == n.op_type and x.outputs[0] == input_blob:
+                input_size = x.attrs['hidden_size']
+
+        model_helper = ModelHelper()
+        name = dummy_name()
+
+        # input and recurrence biases are squashed together in onnx but not in caffe2
+        model_helper.param_init_net.Slice(B, name + "_bias_i2h",   starts=[0],             ends=[4*hidden_size])
+        model_helper.param_init_net.Slice(B, name + "_bias_gates", starts=[4*hidden_size], ends=[8*hidden_size])
+
+        # caffe2 has a different order from onnx. We need to rearrange
+        #   i o f c -> i f o c
+        def reform(name_from, name_to, extra_starts, extra_ends):
+            for start, suffix in enumerate(("_i", "_o", "_f", "_c")):
+                model_helper.param_init_net.Slice(
+                    name_from,
+                    name_from + suffix,
+                    starts = [   start  * hidden_size] + extra_starts,
+                    ends =   [(1+start) * hidden_size] + extra_ends)
+            model_helper.param_init_net.Concat(
+                [name_from + suffix for suffix in ("_i", "_f", "_o", "_c")],
+                [name_to, dummy_name()], axis=0)
+        reform(W, name + "/i2h_w", [0], [input_size])
+        reform(R, name + "/gates_t_w", [0], [hidden_size])
+        reform(name + "_bias_i2h", name + "/i2h_b", [], [])
+        reform(name + "_bias_gates", name + "/gates_t_b", [], [])
+
+        hidden_t_all, hidden_t_last, _, _ = rnn_cell.LSTM(
+            model_helper,
+            input_blob,
+            sequence_lens,
+            [initial_h, initial_c],
+            input_size,
+            hidden_size,
+            name,
+            forward_only=True,
+        )
+
+        model_helper.net = model_helper.net.Clone(
+            "dummy-clone",
+            blob_remap={ hidden_t_all: n.outputs[0], hidden_t_last: n.outputs[1] }
+        )
+
+        init_ops = [ op for op in model_helper.InitProto().op if 'Fill' not in op.type ]
```

Comment:
it's a little bit sketchy. Maybe just construct your 'reform' ops in a separate net? Or do you need some ops produced by LSTMCell directly?

Reply:
separate net could work, I'll try that